File: src/main/java/org/kite9/diagram/visualization/planarization/transform/OuterFaceIdentificationTransform.java
package org.kite9.diagram.visualization.planarization.transform;

import java.util.List;
import java.util.Set;

import org.kite9.diagram.common.algorithms.det.UnorderedSet;
import org.kite9.diagram.common.elements.edge.Edge;
import org.kite9.diagram.common.elements.vertex.MultiCornerVertex;
import org.kite9.diagram.common.elements.vertex.Vertex;
import org.kite9.diagram.model.Diagram;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.visualization.planarization.EdgeMapping;
import org.kite9.diagram.visualization.planarization.Face;
import org.kite9.diagram.visualization.planarization.Planarization;
import org.kite9.diagram.visualization.planarization.mgt.MGTPlanarization;
import org.kite9.framework.common.Kite9ProcessingException;
import org.kite9.framework.logging.Kite9Log;
import org.kite9.framework.logging.Logable;
import org.kite9.framework.logging.LogicException;

/**
 * Works out which faces are outer faces, and identfies them as such.
 */
public class OuterFaceIdentificationTransform implements PlanarizationTransform, Logable {

	private Kite9Log log = new Kite9Log(this);
	
	@Override
	public void transform(Planarization pln) {
		Diagram d = pln.getDiagram();
		EdgeMapping em = pln.getEdgeMappings().get(d);
		Face outerFace = getDiagramOuterFace(em, pln);
		Set<Face> done = new UnorderedSet<Face>(pln.getFaces().size()*2);
		handleOuterFace(outerFace, null, pln, 0, done);
	}
	
	private Face getDiagramOuterFace(EdgeMapping em, Planarization pln) {
		Vertex v = ((MGTPlanarization)pln).getVertexOrder().get(0);
		List<Face> faces = pln.getVertexFaceMap().get(v);
		for (Face face : faces) {
			if ((face.getContainedFaces().size()==0) && isAntiClockwise(face)){
				return face;
			}
		}
		
		throw new LogicException("Couldn't find the outer face");
	}
	
	private boolean isAntiClockwise(Face f) {
		for (int i = 0; i < f.size(); i++) {
			Vertex v = f.getCorner(i);
			Edge e = f.getBoundary(i);
			MultiCornerVertex cv = (MultiCornerVertex) v;
			if ((MultiCornerVertex.isMin(cv.getXOrdinal())) && (MultiCornerVertex.isMin(cv.getYOrdinal()))) {
				return e.getDrawDirectionFrom(v) == Direction.DOWN;
			}
		} 
		
		throw new Kite9ProcessingException();
	}

	private void handleOuterFace(Face outerFace, Face inside, Planarization pln, int level, Set<Face> done) {
		if (done.contains(outerFace)) {
			return;
		}
		
		done.add(outerFace);
		log.send(level(level)+"Outer face: "+outerFace.getId());
		outerFace.setOuterFace(true);
		outerFace.setContainedBy(inside);
		outerFace.getContainedFaces().clear();
		
		traverseInnerFaces(outerFace, pln, level+1, done);
	}

	private void traverseInnerFaces(Face face, Planarization pln, int level, Set<Face> done) {
		for (int i = 0; i < face.size(); i++) {
			Edge e = face.getBoundary(i);
			List<Face> faces = pln.getEdgeFaceMap().get(e);
			for (Face f : faces) {
				if (f != face) {
					handleInnerFace(f, pln, level, done);
				}
			}
		}
	}


	private String level(int level) {
		StringBuilder sb = new StringBuilder(level);
		for (int i = 0; i < level; i++) {
			sb.append(" ");
		}
		return sb.toString();
	}

	private void handleInnerFace(Face innerFace, Planarization pln, int level, Set<Face> done) {
		if (done.contains(innerFace)) {
			return;
		}
		
		done.add(innerFace);
		log.send(level(level)+"Inner face: "+innerFace.getId());
		for (Face f : innerFace.getContainedFaces()) {
			handleOuterFace(f, innerFace, pln, level+1, done);
		}
		
		traverseInnerFaces(innerFace, pln, level, done);
	}

	@Override
	public String getPrefix() {
		return "OFIT";
	}

	@Override
	public boolean isLoggingEnabled() {
		return true;
	}
}

File: src/main/kotlin/org/kite9/diagram/visualization/planarization/transform/OuterFaceIdentificationTransform.kt
package org.kite9.diagram.visualization.planarization.transform

import org.kite9.diagram.common.algorithms.det.UnorderedSet
import org.kite9.diagram.common.elements.edge.Edge
import org.kite9.diagram.common.elements.vertex.MultiCornerVertex
import org.kite9.diagram.common.elements.vertex.MultiCornerVertex.Companion.isMin
import org.kite9.diagram.logging.Kite9Log
import org.kite9.diagram.logging.Logable
import org.kite9.diagram.logging.LogicException
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.visualization.planarization.EdgeMapping
import org.kite9.diagram.visualization.planarization.Face
import org.kite9.diagram.visualization.planarization.Planarization
import org.kite9.diagram.visualization.planarization.mgt.MGTPlanarization

/**
 * Works out which faces are outer faces, and identfies them as such.
 */
class OuterFaceIdentificationTransform : PlanarizationTransform, Logable {

    private val log = Kite9Log(this)

    override fun transform(pln: Planarization) {
        val d = pln.diagram
        val em = pln.edgeMappings[d]
        val outerFace = getDiagramOuterFace(em, pln)
        val done: MutableSet<Face> = UnorderedSet(pln.faces.size * 2)
        handleOuterFace(outerFace, null, pln, 0, done)
    }

    private fun getDiagramOuterFace(em: EdgeMapping?, pln: Planarization): Face {
        val v = (pln as MGTPlanarization).vertexOrder[0]
        val faces = pln.vertexFaceMap[v]
        for (face in faces!!) {
            if (face.containedFaces.size == 0 && isAntiClockwise(face)) {
                return face
            }
        }
        throw LogicException("Couldn't find the outer face")
    }

    private fun isAntiClockwise(f: Face): Boolean {
        for (i in 0 until f.size()) {
            val v = f.getCorner(i)
            val e: Edge = f.getBoundary(i)
            val cv = v as MultiCornerVertex
            if (isMin(cv.xOrdinal) && isMin(cv.yOrdinal)) {
                return e.getDrawDirectionFrom(v) === Direction.DOWN
            }
        }
        throw LogicException("outer face not anti-clockwise")
    }

    private fun handleOuterFace(
        outerFace: Face,
        inside: Face?,
        pln: Planarization,
        level: Int,
        done: MutableSet<Face>
    ) {
        if (done.contains(outerFace)) {
            return
        }
        done.add(outerFace)
        log.send(level(level) + "Outer face: " + outerFace.getID())
        outerFace.isOuterFace = true
        outerFace.containedBy = inside
        outerFace.containedFaces.clear()
        traverseInnerFaces(outerFace, pln, level + 1, done)
    }

    private fun traverseInnerFaces(face: Face, pln: Planarization, level: Int, done: MutableSet<Face>) {
        for (i in 0 until face.size()) {
            val e: Edge = face.getBoundary(i)
            val faces = pln.edgeFaceMap[e]
            for (f in faces!!) {
                if (f !== face) {
                    handleInnerFace(f, pln, level, done)
                }
            }
        }
    }

    private fun level(level: Int): String {
        val sb = StringBuilder(level)
        for (i in 0 until level) {
            sb.append(" ")
        }
        return sb.toString()
    }

    private fun handleInnerFace(innerFace: Face, pln: Planarization, level: Int, done: MutableSet<Face>) {
        if (done.contains(innerFace)) {
            return
        }
        done.add(innerFace)
        log.send(level(level) + "Inner face: " + innerFace.getID())
        for (f in innerFace.containedFaces) {
            handleOuterFace(f, innerFace, pln, level + 1, done)
        }
        traverseInnerFaces(innerFace, pln, level, done)
    }

    override val prefix: String
        get() = "OFIT"

    override val isLoggingEnabled: Boolean
        get() = true
}